packages/actor-scraper/cheerio-scraper/src/internals/request_queue_experimental.ts

```diff
@@ -0,0 +1,186 @@
+        const uid = requestLike.uniqueKey ? requestLike.uniqueKey : requestLike.url;
+        // Faster then negative condition
+        if (this.set.has(uid)) {
+            return;
+        }
+        this.set.add(uid);
+        this.requests.enqueue(this._stripRequest(requestLike));
+    }
+
+    addRequests(requests: []) {
+        for (const request of requests) {
+            this.addRequest(request);
+        }
+    }
+
+    markRequestHandled() {
+        this.handled += 1;
+    }
+
+    handledCount() {
+        return this.handled;
+    }
+
+    isEmpty() {
+        return this.requests.size() === 0;
+    }
+
+    isFinished() {
+        return this.isEmpty();
+    }
+
+    fetchNextRequest(): any {
+        const baseRequest = this.requests.dequeue();
+        return new Request(baseRequest as any as Request);
+    }
+
+    reclaimRequest(request: Request<Dictionary>) {
+        this.requests.enqueue(this._stripRequest(request));
+    }
+
+    async persistQueueState() {
+        const promises = [
+            this.persistRequests(),
+            this.persistSet(),
+        ];
+
+        await Promise.all(promises);
+    }
+
+    async persistSet() {
+        await this._saveInChunks(Array.from(this.set), this.setKey, this.maxBatchSet);
+    }
+
+    async persistRequests() {
+        await this._saveInChunks(this.requests.toArray(), this.queueKey, this.maxBatchRequests);
+    }
+
+    async loadState() {
+        await this._loadRequests();
+        await this._loadSet();
+        this.log.info('State Loaded!', { requestCount: this.requests.size() });
+    }
+
+    _stripRequest(requestLike: Request | RequestOptions) {
+        return {
+            url: requestLike.url,
+            // @ts-expect-error It is
+            retryCount: requestLike.retryCount || 0,
+            uniqueKey: requestLike.uniqueKey,
+            userData: requestLike.userData,
+            label: requestLike.label,
+        };
+    }
+
+    async _loadRequests() {
+        const items = await this._loadItemsFromState(this.queueKey);
+
+        this.requests.fromArray(items);
+    }
+
+    async _loadSet() {
+        const setData = await this._loadItemsFromState(this.setKey);
+
+        this.set = new Set(setData);
+    }
+
+    async _loadItemsFromState(keyRoot: string): Promise<any[]> {
+        const storeClient = await Actor.apifyClient.keyValueStore(this.defaultKeyValueStoreId!);
+
+        const allKeys = await storeClient.listKeys();
+        const orderedKeys = allKeys.items
+            .filter(({ key }) => key.includes(keyRoot))
+            .sort(({ key: key1 }, { key: key2 }) => this._sortByChunkOrder(key1, key2));
+
+        let finalItems: any[] = [];
+        for (const { key } of orderedKeys) {
+            const items = await Actor.getValue(key);
+            finalItems = finalItems.concat(items);
+        }
+
+        await Promise.all(orderedKeys.map(({ key }) => {
+            return storeClient.deleteRecord(key);
```

Comment:
This could be dangerous if you crash right after this delete, you loose everything. I would prefer to figure out how to overwrite properly.